### PR TITLE
feat(vrl): add variant enum argument to the is_json function

### PR DESCRIPTION
**Pull request by nabokihms**
This is an implementation of the idea from the linked issue. 
***
Signed-off-by: m.nabokikh <maksim.nabokikh@flant.com>
Close https://github.com/vectordotdev/vector/issues/12763

### CODE REVIEW

File: lib/vrl/stdlib/src/is_json.rs [hunk 1]
```diff
@@ -10,6 +10,40 @@ fn is_json(value: Value) -> Resolved {
...
+            for c in bytes {
+                return match c {
+                    // Search for the first non whitespace char
+                    b' ' | b'\n' | b'\t' | b'\r' => continue,
+                    b'{' => Ok(value!(variant.as_ref() == b"object")),
+                    b'[' => Ok(value!(variant.as_ref() == b"array")),
+                    b't' | b'f' => Ok(value!(variant.as_ref() == b"bool")),
+                    b'-' | b'0'..=b'9' => Ok(value!(variant.as_ref() == b"number")),
+                    b'"' => Ok(value!(variant.as_ref() == b"string")),
+                    _ => Ok(value!(false)),
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/is_json.rs:28]: We'll want to check for `null` as well.

File: lib/vrl/stdlib/src/is_json.rs [hunk 2]
```diff
@@ -53,11 +99,48 @@ impl Function for IsJson {
         mut arguments: ArgumentList,
     ) -> Compiled {
         let value = arguments.required("value");
-        Ok(Box::new(IsJsonFn { value }))
+        let variant = arguments.optional_enum("variant", &variants())?;
+
+        match variant {
+            Some(raw_variant) => {
+                let variant = raw_variant.try_bytes().expect("variant not bytes");
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/is_json.rs:106]: I suspect you picked up this pattern from other function implementations, but we should bubble up errors, even if the compiler _currently_ prevents this panic from triggering.

```suggestion
                let variant = raw_variant.try_bytes()?;
```

File: lib/vrl/stdlib/src/is_json.rs [hunk 3]
```diff
@@ -53,11 +99,48 @@ impl Function for IsJson {
...
+        _ctx: &mut FunctionCompileContext,
+        name: &str,
+        expr: Option<&expression::Expr>,
+    ) -> CompiledArgument {
+        match (name, expr) {
+            ("variant", Some(expr)) => {
+                let variant = expr
+                    .as_enum("variant", variants())?
+                    .try_bytes()
+                    .expect("variant not bytes");
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/is_json.rs:125]:
```suggestion
                    .try_bytes()?;
```

File: website/cue/reference/remap/functions/is_json.cue [hunk 4]
```diff
@@ -13,6 +13,19 @@ remap: functions: is_json: {
 			required:    true
 			type: ["string"]
 		},
+		{
+			name:        "variant"
+			description: "The variant of the JSON type to explicitly check it."
```
Comment (JeanMertz) [inline website/cue/reference/remap/functions/is_json.cue:18]:
```suggestion
			description: "The variant of the JSON type to explicitly check for."
```

**Review by JeanMertz — CHANGES_REQUESTED** [review]
Sweet! Thank you for adding this so quickly after the previous PR. I left a few comments, but otherwise this LGTM.

File: lib/vrl/stdlib/src/is_json.rs (hunk 2 above)
Comment (nabokihms) [reply lib/vrl/stdlib/src/is_json.rs]: Also converted errors to `Box<dyn DiagnosticMessage>` to make it works 
```
.map_err(|e| Box::new(e) as Box<dyn DiagnosticMessage>)
```

**Review by JeanMertz — APPROVED** [review]

**Pull request merged**